optimism/src/mips/interpreter.rs

```diff
@@ -262,6 +265,72 @@ pub trait InterpreterEnv {
     /// manually add the lookups for this memory operation.
     unsafe fn push_memory_access(&mut self, addr: &Self::Variable, value: Self::Variable);
 
+    /// Access the memory address `addr`, adding constraints asserting that the old value was
+    /// `old_value` and that the new value will be `new_value`.
+    ///
+    /// # Safety
+    ///
+    /// Callers of this function must manually update the memory if required, this function will
+    /// only update the access counter.
+    unsafe fn access_memory(
+        &mut self,
+        addr: &Self::Variable,
+        old_value: &Self::Variable,
+        new_value: &Self::Variable,
+    ) {
+        let last_accessed = {
+            let last_accessed_location = self.alloc_scratch();
+            unsafe { self.fetch_memory_access(addr, last_accessed_location) }
+        };
+        let instruction_counter = self.instruction_counter();
+        let elapsed_time = instruction_counter.clone() - last_accessed.clone();
+        let new_accessed = {
+            // Here, we write as if the memory had been written *at the start of the next
+            // instruction*. This ensures that we can't 'time travel' within this
+            // instruction, and claim to read the value that we're about to write!
+            instruction_counter + Self::constant(1)
+        };
+        self.add_lookup(Lookup {
+            numerator: 1,
+            table_id: LookupTable::MemoryLookup,
+            value: vec![addr.clone(), last_accessed, old_value.clone()],
+        });
+        self.add_lookup(Lookup {
+            numerator: -1,
+            table_id: LookupTable::MemoryLookup,
+            value: vec![addr.clone(), new_accessed, new_value.clone()],
+        });
+        self.range_check64(&elapsed_time);
```

Comment:
why should elapsed time be at most 2^64?

Reply:
I'm allowing for at most 2^64 instructions, mostly arbitrarily.